Run decompositions before the quantizer

mcremon-meta opened the pull request:
Summary:
In the current flow, decompositions run in `to_edge()`, long after the quantization process is done. This creates a lot of issues, since we cannot quantize any operations contained in the large operators that the graph tracer can give (e.g. aten.scaled_dot_product_attention, aten.rnn_<tanh, relu>.input, and a few others).
Any models using those will see many fp32 operators in the final graph. Running the decomps earlier solves the problem, but we need to retain a couple operators that we do rely on in the quantizer, namely `aten.linear`, `aten.conv1d` and `aten.conv2d`.

Differential Revision: D66461406

zonglinpeng commented on backends/cadence/aot/compiler.py:
```diff
@@ -59,15 +60,25 @@ def convert_pt2(
-
-    if model_gm_has_SDPA(model_gm):  # pyre-fixme[6]
+    decomp_table = torch.export.default_decompositions()
+    ops_to_keep = [
```
would be nice to leave the same comment inline

zonglinpeng commented on backends/cadence/aot/compiler.py:
```diff
@@ -59,15 +60,25 @@ def convert_pt2(
+        torch.ops.aten.linear.default,
+        torch.ops.aten.matmul.default,
+    ]
+    # pyre-fixme[6]: For 1st argument expected `Dict[typing.Callable[..., typing.Any
```
pyre are disabled in ET but still used internally. We need to sort it out but not in this PR

mcremon-meta committed:
Run decompositions before the quantizer (#7111)

Summary:

In the current flow, decompositions run in `to_edge()`, long after the quantization process is done. This creates a lot of issues, since we cannot quantize any operations contained in the large operators that the graph tracer can give (e.g. aten.scaled_dot_product_attention, aten.rnn_<tanh, relu>.input, and a few others).
Any models using those will see many fp32 operators in the final graph. Running the decomps earlier solves the problem, but we need to retain a couple operators that we do rely on in the quantizer, like `aten.linear`, `aten.conv1d` and `aten.conv2d`.

Reviewed By: zonglinpeng

Differential Revision: D66461406
